freeipa/src/main/java/com/sequenceiq/freeipa/flow/freeipa/upscale/action/FreeIpaUpscaleActions.java

```diff
@@ -0,0 +1,507 @@
+                Stack stack = context.getStack();
+                stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.UPSCALE_IN_PROGRESS, "Adding instances");
+
+                List<CloudInstance> newInstances = buildNewInstances(context.getStack(), getInstanceCountByGroup(variables));
+                if (newInstances.isEmpty()) {
+                    skipAddingNewInstances(context, stack);
+                } else {
+                    addNewInstances(context, stack, newInstances);
+                }
+            }
+
+            private void skipAddingNewInstances(StackContext context, Stack stack) {
+                List<CloudResourceStatus> list = resourceService.getAllAsCloudResourceStatus(stack.getId());
+                UpscaleStackRequest<UpscaleStackResult> request = new UpscaleStackRequest<>(
+                        context.getCloudContext(), context.getCloudCredential(), context.getCloudStack(), ResourceLists.transform(list));
+                UpscaleStackResult result = new UpscaleStackResult(request.getResourceId(), ResourceStatus.CREATED, list);
+                sendEvent(context, result.selector(), result);
+            }
+
+            private void addNewInstances(StackContext context, Stack stack, List<CloudInstance> newInstances) {
+                Stack updatedStack = instanceMetaDataService.saveInstanceAndGetUpdatedStack(stack, newInstances);
+                List<CloudResource> cloudResources = resourceService.findAllByStackId(stack.getId()).stream()
+                        .map(resource -> resourceConverter.convert(resource))
+                        .collect(Collectors.toList());
+                CloudStack updatedCloudStack = cloudStackConverter.convert(updatedStack);
+                UpscaleStackRequest<UpscaleStackResult> request = new UpscaleStackRequest<>(
+                        context.getCloudContext(), context.getCloudCredential(), updatedCloudStack, cloudResources);
+                sendEvent(context, request.selector(), request);
+            }
+
+            private List<CloudInstance> buildNewInstances(Stack stack, int instanceCountByGroup) {
+                long privateId = getFirstValidPrivateId(stack.getInstanceGroups());
+                List<CloudInstance> newInstances = new ArrayList<>();
+                for (InstanceGroup instanceGroup : stack.getInstanceGroups()) {
+                    int remainingInstances = instanceCountByGroup - instanceGroup.getNotDeletedInstanceMetaDataSet().size();
+                    for (long i = 0; i < remainingInstances; ++i) {
+                        newInstances.add(cloudStackConverter.buildInstance(stack, null, instanceGroup,
+                                stack.getStackAuthentication(), privateId++, InstanceStatus.CREATE_REQUESTED));
+                    }
+                }
+                return newInstances;
+            }
+
+            private Long getFirstValidPrivateId(Set<InstanceGroup> instanceGroups) {
```

Comment:
this seems to me to be a duplicate. we might move this method to a separate class. it would make testing easier

Reply:
I agree its a duplicate. Right now, we have cloudbreak core and freeipa versions of InstanceGroup. They don't share common interfaces. This is the same for a bunch of other classes. While it would be nice to refactor this with a shared API, a shared private instance probably isn't too valuable by itself.